File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/BubbleSort.java
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting;

import java.util.Arrays;

public class BubbleSort extends SortingAlgorithm {
    private int lastConditionCount = -1, lastSwapCount = -1;

    @Override
    int[] sort(int[] randomNumbers) {
        int conditionCounter = 0, swapCounter = 0;
        int length = randomNumbers.length;
        int[] sortedNumbers = Arrays.copyOf(randomNumbers, length);

        for (int i = 0; i < length; ++i) {
            for (int j = 0; j < length - i - 1; ++j) {
                ++conditionCounter;
                if (sortedNumbers[j] > sortedNumbers[j + 1]) {
                    ++swapCounter;
                    int swapTempNumber = sortedNumbers[j];
                    sortedNumbers[j] = sortedNumbers[j + 1];
                    sortedNumbers[j + 1] = swapTempNumber;
                }
            }
        }

        lastConditionCount = conditionCounter;
        lastSwapCount = swapCounter;
        return sortedNumbers;
    }

    @Override
    public String getAlgorithmName() {
        return "Bubble Sort";
    }

    @Override
    int getConditionCount() {
        return lastConditionCount;
    }

    @Override
    public int getSwapCount() {
        return lastSwapCount;
    }
}

File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/BubbleSort.kt
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting

import java.util.*

class BubbleSort : SortingAlgorithm() {
    override var conditionCount = -1
        private set
    override var swapCount = -1
        private set

    override val algorithmName: String
        get() = "Bubble Sort"

    override fun sort(randomNumbers: IntArray): IntArray {
        var conditionCounter = 0
        var swapCounter = 0
        val length = randomNumbers.size
        val sortedNumbers = Arrays.copyOf(randomNumbers, length)

        for (i in 0 until length) {
            for (j in 0 until length - i - 1) {
                ++conditionCounter
                if (sortedNumbers[j] > sortedNumbers[j + 1]) {
                    ++swapCounter
                    val swapTempNumber = sortedNumbers[j]
                    sortedNumbers[j] = sortedNumbers[j + 1]
                    sortedNumbers[j + 1] = swapTempNumber
                }
            }
        }

        conditionCount = conditionCounter
        swapCount = swapCounter
        return sortedNumbers
    }
}
